### PR TITLE
Apply missing patches from (the now unused) internal repo

### PR DESCRIPTION
**Pull request by hunterwerlla**
*Issue #, if available:*
#126, #124, #123, #106, #127
*Description of changes:*
Pull in the existing patches that are already completed but not part of this repo. From now on, this will be the only repo for this project.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 1]
```diff
@@ -163,36 +168,54 @@ export class TaskOperations {
 
         let success = true;
         let environment: Beanstalk.EnvironmentDescription;
+
+        // delay the event poll by a random amount, up to 5 seconds, so that if multiple
+        // deployments run in parallel they don't all start querying at the same time and
+        // potentially trigger throttling from the outset
+        const initialStartDelay = Math.floor(Math.random() * (randomJitterUpperLimit)) + 1;
```
Comment (mpiroc) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: Consider adding an issue to the backlog to use a throttler here instead of a random delay.

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 2]
```diff
@@ -163,36 +168,54 @@ export class TaskOperations {
...
+                requestEvents.StartTime = lastPrintedEventDate;
+                const responseEvent = await this.beanstalkClient.describeEvents(requestEvents).promise();
 
-            if (responseEvent.Events.length > 0) {
-                for (let i = responseEvent.Events.length - 1; i >= 0; i--) {
-                    const event = responseEvent.Events[i];
-                    if (event.EventDate <= lastPrintedEventDate) {
-                        continue;
-                    }
+                if (responseEvent.Events.length > 0) {
```
Comment (mpiroc) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: 1. `if (responseEvent.Events.length > 0)` is redundant, remove this check.
2. This whole loop can be simplified:

```typescript
for (const event of responseEvent.Events.filter(e => e.EventDate <= lastPrintedEventDate)) {
    console.log(event.EventDate + '   ' + event.Severity + '   ' + event.Message)

    if (event.Message === 'Failed to deploy application.') {
        success = false
    }
}
```

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 3]
```diff
@@ -163,36 +168,54 @@ export class TaskOperations {
...
+                        }
                     }
-                }
 
-                lastPrintedEventDate = responseEvent.Events[0].EventDate;
+                    lastPrintedEventDate = responseEvent.Events[0].EventDate;
+                }
+            } catch (err) {
+                // if we are still encountering throttles, increase the poll delay some more
+                if (err.code === 'Throttling') {
```
Comment (mpiroc) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: I ***highly*** recommend enabling (at minimum) the `no-unsafe-any` tslint rule, which will prevent you from trying to access members of `any`. Without this rule enabled, the type system is almost worthless and type unsafety will propogate throughout the codebase.

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 4]
```diff
@@ -163,36 +168,54 @@ export class TaskOperations {
...
+                    eventPollDelay += Math.floor(Math.random() * (randomJitterUpperLimit)) + 1;
```
Comment (mpiroc) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: This logic is duplicated above. Consider moving it to a helper function.

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts [hunk 5]
```diff
@@ -15,13 +15,18 @@ export class TaskParameters extends AWSTaskParametersBase {
     public static readonly fileSource: string = 'file';
```
Comment (mpiroc) [inline Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: 1. Nit: Suggest to remove the `: string` annotations, as they prevent the TS compiler from inferring the more specific types `urlSource[: 'url'] = 'url'`, etc. The string literal types add extra type safety, and provide extra context when you mouse-over them in your favorite editor/IDE.
2. Rather than using `static readonly` class fields, consider exporting these as top-level constants:

  ```typescript
  export const urlSource = 'url'
  export const s3Source = 's3'
  // etc.
  ```

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts [hunk 6]
```diff
@@ -83,7 +90,7 @@ export class TaskParameters extends AWSTaskParametersBase {
             }
 
             this.templateParametersSource = tl.getInput('templateParametersSource', true);
-            switch (this.templateParametersSource) {
+			switch (this.templateParametersSource) {
```
Comment (mpiroc) [inline Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: Kill tabs with fire.

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts [hunk 7]
```diff
@@ -98,17 +105,18 @@ export class TaskParameters extends AWSTaskParametersBase {
...
                 case TaskParameters.loadTemplateParametersInline: {
                     this.templateParameters = tl.getInput('templateParameters', true);
-                }
                     break;
+                }
 
                 default:
                     throw new Error(`Unrecognized template parameters source: ${this.templateParametersSource}`);
             }
+
```
Comment (mpiroc) [inline Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: Nit: Multiple consecutive empty lines.

File: Tasks/CloudFormationExecuteChangeSet/helpers/ExecuteChangeSetTaskParameters.ts [hunk 8]
```diff
@@ -10,16 +10,25 @@ import tl = require('vsts-task-lib/task');
 import { AWSTaskParametersBase } from 'sdkutils/awsTaskParametersBase';
 
 export class TaskParameters extends AWSTaskParametersBase {
+
+    public static readonly ignoreStackOutputs: string = 'ignore';
```
Comment (mpiroc) [inline Tasks/CloudFormationExecuteChangeSet/helpers/ExecuteChangeSetTaskParameters.ts]: See above--consider converting these to top-level exported constants.

File: Tasks/Common/cloudformationutils/cloudformationutils.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  */
+
+import tl = require('vsts-task-lib/task');
+import CloudFormation = require('aws-sdk/clients/cloudformation');
+
+export class CloudFormationUtils {
+    // Retrieves the declared outputs of the stack and posts as either individual variables,
+    // using the output member key as variable name, or a json-formatted blob with a variable
+    // name of the stack name suffixed with 'Outputs'
+    public static async captureStackOutputs(cloudFormationClient: CloudFormation,
```
Comment (mpiroc) [inline Tasks/Common/cloudformationutils/cloudformationutils.ts]: Throughout this PR, consider replacing static class methods with top-level exported functions.

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 1 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: Filed https://github.com/aws/aws-vsts-tools/issues/171

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 2 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: sure we can do that refactor, these kinds of things need to be fixed eventually

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 3 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: we have that in the other PR

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 4 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: I will add that when this module is refactored

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts (hunk 5 above)
Comment (hunterwerlla) [reply Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: when this module is refactored we can look into that

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts (hunk 6 above)
Comment (hunterwerlla) [reply Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: fixed

File: Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts (hunk 7 above)
Comment (hunterwerlla) [reply Tasks/CloudFormationCreateOrUpdateStack/helpers/CreateOrUpdateStackTaskParameters.ts]: fixed

File: Tasks/CloudFormationExecuteChangeSet/helpers/ExecuteChangeSetTaskParameters.ts (hunk 8 above)
Comment (hunterwerlla) [reply Tasks/CloudFormationExecuteChangeSet/helpers/ExecuteChangeSetTaskParameters.ts]: We can think about that when this is refactored, please put that feedback on this PR: https://github.com/aws/aws-vsts-tools/pull/168

File: Tasks/Common/cloudformationutils/cloudformationutils.ts (hunk 9 above)
Comment (hunterwerlla) [reply Tasks/Common/cloudformationutils/cloudformationutils.ts]: We can do that when it's refactored please put comment on https://github.com/aws/aws-vsts-tools/pull/168

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 10]
```diff
@@ -191,22 +191,15 @@ export class TaskOperations {
...
+                for (const event of responseEvent.Events.filter(e => e.EventDate <= lastPrintedEventDate)) {
+                    console.log(event.EventDate + '   ' + event.Severity + '   ' + event.Message)
 
-                    lastPrintedEventDate = responseEvent.Events[0].EventDate;
+                    if (event.Message === 'Failed to deploy application.') {
+                        success = false
+                    }
                 }
+
+                lastPrintedEventDate = responseEvent.Events[0].EventDate;
```
Comment (awschristou) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: `responseEvent.Events` could be empty

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts [hunk 11]
```diff
@@ -191,22 +191,15 @@ export class TaskOperations {
...
-                            continue;
-                        }
-
-                        console.log(event.EventDate + '   ' + event.Severity + '   ' + event.Message);
-
-                        if (event.Message === 'Failed to deploy application.') {
-                            success = false;
-                        }
-                    }
+                for (const event of responseEvent.Events.filter(e => e.EventDate <= lastPrintedEventDate)) {
```
Comment (awschristou) [inline Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: 1. Logs used to output in reverse order. Is this change intended?
2. the filter logic appears to be reversed (old code filtered out, new code filters in)

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 10 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: This error was already here (yay)

File: Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts (hunk 11 above)
Comment (hunterwerlla) [reply Tasks/BeanstalkDeployApplication/helpers/DeployApplicationTaskOperations.ts]: 1.  Hm, no: https://docs.aws.amazon.com/elasticbeanstalk/latest/api/API_DescribeEvents.html aparently it returns most recent
2. @mpiroc :P
